File: locust/auth_lifecycle.py
import time
from locust import HttpUser, between, task
from common import PASSWORD, ProductionTrafficShape, auth_headers


class AuthLifecycle(HttpUser):
    wait_time = between(1, 3)

    @task
    def lifecycle(self):
        email = f"perf-{id(self)}-{time.time_ns()}@example.test"
        payload = {"email": email, "password": PASSWORD, "firstName": "Load", "lastName": "User"}
        with self.client.post("/api/auth/register", json=payload, headers=auth_headers(), name="POST /api/auth/register", catch_response=True) as response:
            if response.status_code != 201: response.failure(f"expected 201, got {response.status_code}")
        with self.client.post("/api/auth/login", json={"email": email, "password": PASSWORD}, headers=auth_headers(), name="POST /api/auth/login", catch_response=True) as response:
            if response.status_code != 200: response.failure(f"expected 200, got {response.status_code}")
            data = response.json() if response.status_code == 200 else {}
        refresh = data.get("tokens", {}).get("refreshToken")
        with self.client.post("/api/auth/refresh", json={"email": email, "password": PASSWORD, "refreshToken": refresh}, headers=auth_headers(), name="POST /api/auth/refresh", catch_response=True) as response:
            if response.status_code != 200: response.failure(f"expected 200, got {response.status_code}")
            elif response.json().get("tokens", {}).get("refreshToken"): refresh = response.json()["tokens"]["refreshToken"]
        with self.client.post("/api/auth/logout", json={"email": email, "password": PASSWORD, "refreshToken": refresh}, headers=auth_headers(), name="POST /api/auth/logout", catch_response=True) as response:
            if response.status_code != 200: response.failure(f"expected 200, got {response.status_code}")


class AuthShape(ProductionTrafficShape):
    peak_rate = 22
    production_users = 201

File: k6/auth-lifecycle.js
import http from 'k6/http';
import { check } from 'k6';
import { BASE_URL, PASSWORD, jsonParams, parseJson, scenarioOptions, think } from './common.js';

export const options = scenarioOptions(22, 201, 402);

export default function () {
  const email = `perf-${__VU}-${__ITER}-${Date.now()}@example.test`;
  let response = http.post(`${BASE_URL}/api/auth/register`, JSON.stringify({email, password: PASSWORD, firstName: 'Load', lastName: 'User'}), jsonParams());
  check(response, {'register 201': (r) => r.status === 201});
  think(1, 3);
  response = http.post(`${BASE_URL}/api/auth/login`, JSON.stringify({email, password: PASSWORD}), jsonParams());
  check(response, {'login 200': (r) => r.status === 200});
  let token = parseJson(response).tokens?.refreshToken;
  think(1, 3);
  response = http.post(`${BASE_URL}/api/auth/refresh`, JSON.stringify({email, password: PASSWORD, refreshToken: token}), jsonParams());
  check(response, {'refresh 200': (r) => r.status === 200});
  token = parseJson(response).tokens?.refreshToken || token;
  think(1, 3);
  response = http.post(`${BASE_URL}/api/auth/logout`, JSON.stringify({email, password: PASSWORD, refreshToken: token}), jsonParams());
  check(response, {'logout 200': (r) => r.status === 200});
}
